modules/apps/depot/depot-service/src/main/java/com/liferay/depot/service/impl/DepotEntryPinServiceImpl.java

```diff
@@ -0,0 +1,144 @@
+import org.osgi.service.component.annotations.ReferencePolicyOption;
+
+/**
+ * @author Brian Wing Shun Chan
+ */
+@Component(
+	property = {
+		"json.web.service.context.name=depot",
+		"json.web.service.context.path=DepotEntryPin"
+	},
+	service = AopService.class
+)
+public class DepotEntryPinServiceImpl extends DepotEntryPinServiceBaseImpl {
+
+	@Override
+	public DepotEntryPin addDepotEntryPin(long userId, long depotEntryId)
+		throws PortalException {
+
+		_depotEntryModelResourcePermission.check(
+			getPermissionChecker(), depotEntryId, ActionKeys.UPDATE);
+
+		return depotEntryPinLocalService.addDepotEntryPin(userId, depotEntryId);
+	}
+
+	@Override
+	public DepotEntryPin deleteDepotEntryPin(long userId, long depotEntryId)
+		throws PortalException {
+
+		_depotEntryModelResourcePermission.check(
+			getPermissionChecker(), depotEntryId, ActionKeys.UPDATE);
+
+		return depotEntryPinLocalService.deleteDepotEntryPin(
+			userId, depotEntryId);
+	}
+
+	@Override
+	public List<DepotEntryPin> getDepotEntryDepotEntryPins(
+			long depotEntryId, int start, int end)
+		throws PortalException {
+
+		_depotEntryModelResourcePermission.check(
+			getPermissionChecker(), depotEntryId, ActionKeys.VIEW);
+
+		return depotEntryPinLocalService.getDepotEntryDepotEntryPins(
+			depotEntryId, start, end);
+	}
+
+	@Override
+	public int getDepotEntryDepotEntryPinsCound(
+			long depotEntryId, int start, int end)
+		throws PortalException {
+
+		_depotEntryModelResourcePermission.check(
+			getPermissionChecker(), depotEntryId, ActionKeys.VIEW);
+
+		return depotEntryPinLocalService.getDepotEntryDepotEntryPinsCount(
+			depotEntryId);
+	}
+
+	@Override
+	public List<DepotEntryPin> getUserDepotEntryPins(
+			long userId, int start, int end)
+		throws PortalException {
+
+		List<DepotEntryPin> depotEntryPins = new ArrayList<>();
+
+		while (depotEntryPins.size() < end) {
+			List<DepotEntryPin> userDepotEntryPins =
+				depotEntryPinPersistence.findByUserId(userId, start, end);
+
+			if (userDepotEntryPins.isEmpty()) {
+				break;
+			}
+
+			for (DepotEntryPin userDepotEntryPin : userDepotEntryPins) {
+				if (_depotEntryModelResourcePermission.contains(
+						getPermissionChecker(),
+						userDepotEntryPin.getDepotEntryId(), ActionKeys.VIEW)) {
+
+					depotEntryPins.add(userDepotEntryPin);
+				}
+
+				start = start + end;
+
+				end = start + end;
+			}
+		}
+
+		return depotEntryPins;
+	}
+
+	@Override
+	public int getUserDepotEntryPinsCount(long userId) throws PortalException {
+		int count = 0;
+
+		List<DepotEntryPin> userDepotEntryPins =
+			depotEntryPinPersistence.findByUserId(userId);
+
+		if (userDepotEntryPins.isEmpty()) {
+			return count;
+		}
+
+		for (DepotEntryPin userDepotEntryPin : userDepotEntryPins) {
+			if (_depotEntryModelResourcePermission.contains(
+					getPermissionChecker(), userDepotEntryPin.getDepotEntryId(),
+					ActionKeys.VIEW)) {
+
+				count++;
+			}
+		}
+
+		return count;
+	}
```

Comment:
This looks bad. I'd prefer not to filter the pins. We could, for example, add a listener when the DEPOT_MEMBER role is removed from an user and delete the pins then instead of filtering like this.